README.md
# saturation

Calculate sequencing saturation in SAW

## 程序版本

本文档对应程序版本为 **v2.3.5**

## 运行环境要求

linux系统环境

python3要求安装,且在环境变量PATH中,需要安装python库包括:

* numpy
* pandas
* matplotlib
* scipy

## 输入文件要求

* 原始基因表达文件. 名称为: raw_barcode_gene_exp.txt **兼容格式有表头/无表头,表头需按特定格式,五列数据以空格分隔**
* 组织区域坐标文件. 名称为: SN.gem.gz或SN.gef **前者要求格式有注释行,有表头,四列数据以tab分隔**
* 比对统计文件. 名称为: SN.barcodeMap.stat **列以tab分隔,包含以 total_reads: 开头的行**
* 注释统计文件. 名称为: SN.summary.stat **要求有表头,列以tab分隔**

## 输出文件说明

* 测序饱和度统计文件.
  文件共九列,以tab分隔;第一行为表头,示例如下:
  | 抽样比例 | bin1 total reads | bin1 饱和度值 | bin1 基因数中值 | bin1 uniq reads | bin200 total reads | bin200 饱和度值 | bin200 基因数中值 | bin200 uniq reads |
  | -------- | ---------------- | ------------- | --------------- | --------------- | ------------------ | --------------- | ----------------- | -------------- |
  | #sample  | bar_x            | bar_y1        | bar_y2          | bar_umi         | bin_x              | bin_y1          | bin_y2            | bin_umi        |
  | 0.05     | 107416           | 0.5978346     | 1               | 43199           | 107416             | 0.5981511       | 13                | 51             |
  | 0.1      | 214832           | 0.718608      | 1               | 60452           | 214832             | 0.7188687       | 19                | 69             |
  | 0.2      | 429665           | 0.8042382     | 1               | 84112           | 429665             | 0.8044639       | 29                | 91             |
  | 0.3      | 644498           | 0.8399685     | 1               | 103140          | 644498             | 0.8402043       | 37                | 109            |
  | 0.4      | 859330           | 0.8606996     | 1               | 119705          | 859330             | 0.8609568       | 43                | 124            |
  | 0.5      | 1074163          | 0.8747304     | 1               | 134560          | 1074163            | 0.8749957       | 49                | 137            |
  | 0.6      | 1288996          | 0.8849469     | 1               | 148303          | 1288996            | 0.8852138       | 55                | 148            |
  | 0.7      | 1503828          | 0.8928634     | 1               | 161115          | 1503828            | 0.8931374       | 58                | 160            |
  | 0.8      | 1718661          | 0.8992623     | 1               | 173134          | 1718661            | 0.8995398       | 62                | 170            |
  | 0.9      | 1933494          | 0.9044445     | 1               | 184756          | 1933494            | 0.9047227       | 64                | 180            |
  | 1        | 2148327          | 0.9088984     | 1               | 195716          | 2148327            | 0.9091828       | 66                | 189            |
* 饱和度图片. 所有图片的x轴都是相同的,读取统计文件第二列(bin1 total reads),并按比例还原回total reads,即最后一个采样点的x轴坐标对应total reads(即可视化页面的total reads)
  * bin1图片. 共两幅图,分别以第三列(bin1 饱和度值),第四列(bin1 基因数中值)作为y轴
  * bin200图. 共三幅图,前两幅分别以第七列(bin200 饱和度值),第八列(bin200 基因数中值)作为y轴; 第三幅图以最后一列(bin200 uniq reads)作为y轴,并对该统计结果做对数拟合,将拟合公式输出到统计图上,如果最大取样量下的值小于设定的阈值5000,并且拟合曲线的R2 大于等于0.9,就画出拟合曲线的延长线,一直延长到阈值处,并在图上画出阈值处的坐标值

## 程序处理逻辑

饱和度值计算公式为 *1-(uniq reads/total reads)*

* 准备. 
  * 读取比对统计文件,获取total reads
  * 读取注释统计文件,获取anno reads
  * 计算比例anno reads/total reads,用于画图时还原x轴最大坐标为total reads
* 读tissue数据文件. 
  * 将所有unique 坐标提取出来,注意这里的坐标需要加上offset
  * 将坐标取bin,得到bin200的unique 坐标
  * 随机采样5%的bin200 unique坐标,并还原回bin1的坐标,用于过滤表达矩阵中的数据
* 读表达矩阵文件.
  * 使用tissue下的采样后的bin1坐标过滤数据,构建(x,y,gene,mid)的列表
  * 同时累加所有count值,得到anno reads
* 饱和度计算.
  * 将前一列表顺序打乱
  * 按照{0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0}的采样间隔,顺序处理数据
  * 对每一个采样点,分别计算bin1/bin200下的total reads/饱和度值/基因数中值等,输出统计结果文件
    * **饱和度值**. 计算所有bin下的uniq reads(即gene,mid都为uniq)和total reads,使用公式1-(uniq reads/total reads)计算得出; bin1 bin200算法相同
    * **基因数中值**. 计算每个bin下的uniq gene个数,取中值; bin1 bin200算法相同
    * **uniq reads**. bin1的uniq reads为所有bin1下的uniq reads(即gene,mid都为uniq); bin200的uniq reads为所有bin200下的uniq reads(bin1的x y坐标,gene和mid都为uniq)
* 绘图. 分别绘制bin1/bin200的统计图


CHANGELOG.md
# CHANGELOG

## v2.3.5(2022-10-20)

* 格式化错误码

## v2.3.4(2022-10-14)

* 兼容exp文件无表头情况

## v2.3.3(2022-09-13)

* 适配输入exp文件有表头的格式
* 将输出bin200的图片形式由2x2改成1x3

## v2.3.2(2022-08-18)

* 输入统计文件检查. 如果map reads为0或map reads小于anno reads,则程序抛出异常
* 修复当数据量很小导致的统计除0报错问题,并进一步修改逻辑,如果采样之后数据很少则采用全部数据进行计算

## v2.3.1(2022-05-09)

* 将图片x轴坐标与total reads对应

## v2.3.0(2022-04-11)

* 使用C++重构原饱和度计算相关python代码,以降低内存并加速计算过程
* 解决集群画图报错问题
* 更新bin200图片输出格式,由1x3改为2x2,并更新图片字体大小等属性
* 格式化饱和度统计文件格式为tsv

## v2.2.1(2022-01-27)

* 支持注释统计文件以列表形式输入

## v2.2.0(2022-01-24)

* 支持输入tissue 基因表达矩阵文件格式为gef格式



script/build.sh

# change the libPath to your specific path
libPath="/root/lib"
python3="python3"


cmakePath="$libPath/cmake-3.17.2"
gccPath="$libPath/gcc-9.1.0"
CLI11Path="$libPath/CLI11-1.9.0"
hdf5Path="$libPath/hdf5-1.12.1"
libxPath="$libPath/libx-1.1"


binPath="/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/bin"
export PATH="$gccPath/bin:$cmakePath/bin:$binPath"

#export LD_LIBRARY_PATH="$gccPath/lib64:$LD_LIBRARY_PATH"
export LD_LIBRARY_PATH=$gccPath/lib64:$hdf5Path/lib
export LIBRARY_PATH=$LD_LIBRARY_PATH

export C_INCLUDE_PATH="$libxPath/include:$hdf5Path/include:$CLI11Path/include:$gccPath/include:$C_INCLUDE_PATH"
export CPLUS_INCLUDE_PATH=$C_INCLUDE_PATH

export CC="$gccPath/bin/gcc"
export CXX="$gccPath/bin/g++"

absPath(){
    relativePath=$1
    mkdir -p $relativePath
    if [ ${relativePath:0:1} == "/" ]
    then
        echo $relativePath
        return 0
    fi
    ap="$(cd $relativePath; pwd)"
    echo "$ap"
}

install(){
    libFile="$1"
    tlib="$installPath/lib"
    if [ -d "$tlib" ]
    then
        if [ -e "$libFile" ]
        then
            echo "Installing: $libFile into $tlib"
            cp $libFile $tlib
        else
            echo "[WARN] File not exists: $libFile"
        fi
    fi
}

srcPath="$(cd $(dirname $(dirname $0)); pwd)"
buildPath="build"
buildPath="$(absPath $buildPath)"
mkdir -p $buildPath
cd $buildPath
installPath="$buildPath/install"
mkdir -p $installPath/bin $installPath/lib

timeStart=$(date +%s)

cmake $srcPath -DINSTALL_PATH=$installPath -DHDF5PATH=$hdf5Path/lib

thread=$(grep -c ^processor /proc/cpuinfo)
make -j $thread #VERBOSE=1

if [[ $? == 0 ]];
then
    install $gccPath/lib64/libstdc++.so.6
    install $gccPath/lib64/libgomp.so.1
    install $gccPath/lib64/libgcc_s.so.1
    install $hdf5Path/lib/libhdf5.so.200
    # install $hdf5Path/lib/libhdf5_cpp.so.200

    find $srcPath/src -name "*.py"| xargs -I {} $python3 -c "import py_compile;py_compile.compile('{}',cfile='{}c')"
    cp $srcPath/src/plot.pyc $installPath/bin/

    cd $srcPath
    cp -R $installPath $srcPath
    echo -n "Success! "
else
    echo -n "Fail! "
fi

timeEnd=$(date +%s)
secs=$(($timeEnd - $timeStart))
mins=$(($secs/60))
secs=$(($secs%60))
echo "Cost: $mins Mins $secs Secs"


src/saturation.cpp
/* Copyright (C) BGI-Reasearch - All Rights Reserved
* Unauthorized copying of this file, via any medium is strictly prohibited
* Proprietary and confidential
* Written by STOmics development team P_stomics_dev@genomics.cn, May 2017
*/
#include <iostream>
#include <unordered_set>
#include <set>
#include <unordered_map>
#include <map>
#include <vector>
#include <string>
#include <tuple>
#include <sstream>
#include <filesystem>
using namespace std;
namespace fs = std::filesystem;

// #include "H5Cpp.h"
// using namespace H5;
#include "hdf5.h"

#include "libx/File.hpp"
#include "libx/GZFile.hpp"
#include "libx/String.hpp"
#include "libx/System.hpp"
#include "libx/Timer.hpp"

#include "CLI11.hpp"

typedef unsigned char uint8;
typedef unsigned short uint16;
typedef unsigned int uint32;
typedef unsigned long uint64;

#define DEBUG

static constexpr int BIN_SIZE = 200;
static constexpr float SAMPLE_RATIOS[] = {0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0};
static constexpr float DATA_RATIO = 0.05;
static constexpr int MIN_BIN_THRE = 5;

// Store raw data from gene expression file
struct ExpLine
{
    ExpLine(uint32 cx, uint32 cy, uint32 gene, uint64 mid):
            cx(cx), cy(cy), gene(gene), mid(mid)
    {
    }
    uint32 cx;
    uint32 cy;
    uint32 gene;
    uint64 mid;
};

// Store raw data from tissue.gef
typedef struct exp_t
{
    uint32 x;
    uint32 y;
    uint32 cnt;
} exp_t;

typedef tuple<uint32, uint64> GeneMID;
typedef tuple<uint32, uint32, uint32, uint64> CoorGeneMID;

typedef unordered_map<uint64, map<GeneMID, uint32>> SatStatMap;
typedef unordered_map<uint64, set<CoorGeneMID>> MIDStatMap;

struct Arguments
{
    string expFile;
    string outDir;
    string tissueFile;

    vector<string> annoStatFiles;
    vector<string> bcMapStatFiles;

    string outSatFile;
};
Arguments arguments;

// Bin path, for executing plot.py
fs::path binPath;

static const std::string version = "2.3.5";

void save_error_code(int errcode, string info)
{
    std::ostringstream ostr;
    std::time_t        t =
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    ostr << std::put_time(std::localtime(&t), "[%Y-%m-%d %H:%M:%S] SAW-A8000") << errcode
         << ": " << info << endl;

    // std::cerr<<ostr.str();

    string filename("errcode.log");
    libx::writeFile(ostr.str(), filename, true);
}

bool parseArgs(int argc, char** argv)
{
    libx::Timer timer;
    // Parse the command line parameters.
    CLI::App app{ "Sequencing Saturation and Plotting." };
    app.footer("saturation version: " + version);
    app.get_formatter()->column_width(40);

    // Required parameters
    app.add_option("-I,-i", arguments.expFile,
                   "Input gene expression matrix contains reads count information");
    app.add_option("-O,-o", arguments.outDir, "Output directory");
    app.add_option("--tissue", arguments.tissueFile, 
                   "Gene expression matrix only contians data under the tissue region");
    string annoStatFiles;
    app.add_option("--summary", annoStatFiles, 
        "Annotation summary file, support multi-files seperated by ','");
    string bcMapStatFiles;
    app.add_option("--bcstat", bcMapStatFiles, 
        "Barcode mapping statistics report, support multi-files seperated by ','");

    app.callback([&]() {
        if (arguments.expFile.empty())
        {
            save_error_code(1, "-I is missing.");
            exit(-1);
        }
        if (!fs::exists(arguments.expFile))
        {
            save_error_code(2, "not exists file: " + arguments.expFile);
            exit(-1);
        }
        if (arguments.outDir.empty())
        {
            save_error_code(1, "-O is missing.");
            exit(-1);
        }
        if (!fs::exists(arguments.outDir))
        {
            save_error_code(2, "not exists path: " + arguments.outDir);
            exit(-1);
        }
        if (arguments.tissueFile.empty())
        {
            save_error_code(1, "--tissue is missing.");
            exit(-1);
        }
        if (!fs::exists(arguments.tissueFile))
        {
            save_error_code(2, "not exists file: " + arguments.tissueFile);
            exit(-1);
        }
        if (annoStatFiles.empty())
        {
            save_error_code(1, "--summary is missing.");
            exit(-1);
        }
        if (bcMapStatFiles.empty())
        {
            save_error_code(1, "--bcstat is missing.");
            exit(-1);
        }
    });
        
    try 
    {
        app.parse(argc, argv);
    } 
    catch (const CLI::ParseError &e) 
    {
        save_error_code(1, "invalid parameters.");
        app.exit(e);
        return false;
    }

    arguments.outSatFile = arguments.outDir + "/sequence_saturation.tsv";
    // Check files is exist
    vector<string> temp;
    libx::split(annoStatFiles, ',', temp);
    for (auto& f : temp)
    {
        if (f.empty()) continue;
        if (!fs::exists(f))
        {
            save_error_code(2, "not exists file: " + f);
            exit(-1);
        }
        arguments.annoStatFiles.push_back(f);
    }
    temp.clear();
    libx::split(bcMapStatFiles, ',', temp);
    for (auto& f : temp)
    {
        if (f.empty()) continue;
        if (!fs::exists(f))
        {
            save_error_code(2, "not exists file: " + f);
            exit(-1);
        }
        arguments.bcMapStatFiles.push_back(f);
    }
    if (arguments.annoStatFiles.empty())
    {
        save_error_code(2, "--summary has no valid files.");
        exit(-1);
    }
    if (arguments.bcMapStatFiles.empty())
    {
        save_error_code(2, "--bcstat has no valid files.");
        exit(-1);
    }

    binPath = argv[0];
    binPath = fs::absolute(binPath).parent_path();

    return true;
}

void parseGefTissue(string tissueFile, unordered_set<uint64>& uniqCoors)
{
    vector<exp_t> data;
    uint32 minx, miny;

    try
    {
        hid_t m_file_id;
        hid_t m_dataspace_id;
        hid_t m_dataset_id;
        hsize_t dims[1];
        hid_t attr;

        m_file_id  = H5Fopen(tissueFile.c_str(), H5F_ACC_RDONLY, H5P_DEFAULT);

        char expName[128]={0};
        sprintf(expName, "/geneExp/bin1/expression");
        m_dataset_id = H5Dopen(m_file_id, expName, H5P_DEFAULT);
        m_dataspace_id = H5Dget_space(m_dataset_id);
        H5Sget_simple_extent_dims(m_dataspace_id, dims, NULL);
        auto expLen = dims[0];
#ifdef DEBUG
        cout<<"gef expression lines: "<<expLen<<endl;
#endif

        hid_t memtype = H5Tcreate(H5T_COMPOUND, sizeof(exp_t));
        H5Tinsert(memtype, "x", HOFFSET(exp_t, exp_t::x), H5T_NATIVE_UINT);
        H5Tinsert(memtype, "y", HOFFSET(exp_t, exp_t::y), H5T_NATIVE_UINT);
        H5Tinsert(memtype, "count", HOFFSET(exp_t, exp_t::cnt), H5T_NATIVE_UINT);

        data.resize(expLen);
        H5Dread(m_dataset_id, memtype, H5S_ALL, H5S_ALL, H5P_DEFAULT, &data[0]);

        attr = H5Aopen(m_dataset_id, "minX", H5P_DEFAULT);
        H5Aread(attr, H5T_NATIVE_UINT, &minx);
        attr = H5Aopen(m_dataset_id, "minY", H5P_DEFAULT);
        H5Aread(attr, H5T_NATIVE_UINT, &miny);
#ifdef DEBUG
        cout<<"gef offset: "<<minx<<" "<<miny<<endl;
#endif

        H5Aclose(attr);
        H5Tclose(memtype);
        H5Dclose(m_dataset_id);
        H5Sclose(m_dataspace_id);
        H5Fclose(m_file_id);
    }
    catch (...)
    {
        save_error_code(3, "invalid tissue file: " + tissueFile);
        exit(-1);
    }

    // Extract unique coordinates
    for (auto& t : data)
    {
        uint64 coor = ((uint64)(t.x+minx) << 32) + t.y+miny;
        uniqCoors.insert(coor);
    }
}

// Used for hdf5 library compiled with supporting cpp,
// unforunately, we can't use it in cluster
// void parseGefTissueH5CPP(string tissueFile, unordered_set<uint64>& uniqCoors)
// {
//     vector<exp_t> data;
//     uint32 minx, miny;

//     try
//     {
//         H5File* file = new H5File(tissueFile.c_str(), H5F_ACC_RDONLY);
//         DataSet* dataset = new DataSet(file->openDataSet("/geneExp/bin1/expression"));

//         DataSpace dataspace = dataset->getSpace();
//         hsize_t dims[1];
//         dataspace.getSimpleExtentDims(dims, NULL);

//         Attribute attr = dataset->openAttribute("minX");
//         attr.read(PredType::NATIVE_UINT32, &minx);
//         attr = dataset->openAttribute("minY");
//         attr.read(PredType::NATIVE_UINT32, &miny);

//         CompType mtype(sizeof(exp_t));
//         mtype.insertMember("x", HOFFSET(exp_t, x), PredType::NATIVE_UINT32);
//         mtype.insertMember("y", HOFFSET(exp_t, y), PredType::NATIVE_UINT32);
//         mtype.insertMember("count", HOFFSET(exp_t, cnt), PredType::NATIVE_UINT32);

//         data.resize(dims[0]);
//         dataset->read(&data[0], mtype);
        
//         delete dataset;
//         delete file;
//     }
//     catch (...)
//     {
//         cerr<<"SAW-A80001 invalid tissueFile: "<<tissueFile<<endl;
//         throw std::runtime_error("invalid tissueFile!");
//     }

//     // Extract unique coordinates
//     for (auto& t : data)
//     {
//         uint64 coor = ((uint64)(t.x+minx) << 32) + t.y+miny;
//         uniqCoors.insert(coor);
//     }
// }

void parseGemTissue(string tissueFile, unordered_set<uint64>& uniqCoors)
{
    // Parse header for minx, miny
    uint32 minx = 0, miny = 0;
    auto parseHeader = [&](std::string& line){
        if (line[0] != '#') return false;

        if (libx::startswith(line, "#OffsetX"))
        {
            vector<string> temp;
            libx::split(line, '=', temp);
            minx = std::stoi(temp[1]);
        }
        else if (libx::startswith(line, "#OffsetY"))
        {
            vector<string> temp;
            libx::split(line, '=', temp);
            miny = std::stoi(temp[1]);
        }
        return true;
    };
    if (libx::endswith(tissueFile, ".gz"))
        libx::readGZFile(tissueFile, parseHeader, "!", 0);
    else
        libx::readFile(tissueFile, parseHeader, "!", 0);

    // Parse whole data
    vector<uint64> tempCoors;
    auto parseData = [&](std::string& line){
        // Extract unique coordinates=
        string gene;
        int x, y, cnt;
        libx::split(line, '\t', gene, x, y, cnt);
        uint64 coor = ((uint64)(x+minx) << 32) + y+miny;
        tempCoors.emplace_back(coor);

        return true;
    };
    if (libx::endswith(tissueFile, ".gz"))
        libx::readGZFile(tissueFile, parseData, "#", 1);
    else
        libx::readFile(tissueFile, parseData, "#", 1);

    for (auto& coor : tempCoors)
        uniqCoors.insert(coor);

}

bool extractTissueCoor(string tissueFile, unordered_set<uint64>& tissueSampleCoors)
{
    unordered_set<uint64> tissueRawCoors;
    if (libx::endswith(tissueFile, ".gef"))
        parseGefTissue(tissueFile, tissueRawCoors);
    else if (libx::endswith(tissueFile, ".gem.gz") || libx::endswith(tissueFile, ".gem"))
        parseGemTissue(tissueFile, tissueRawCoors);
    
    // Sampleing uniq corrdinates with bin200
    unordered_set<uint64> binCoors;
    for (auto& coor : tissueRawCoors)
    {
        uint32 y = coor & 0xFFFFFFFF;
        uint32 x = (coor >> 32) & 0xFFFFFFFF;
        uint64 binCoor = ((uint64)x/BIN_SIZE << 32) + y/BIN_SIZE;
        binCoors.insert(binCoor);
    }

    float sampleRatio = DATA_RATIO;
    int sampleNum = binCoors.size() * sampleRatio;
    if (sampleNum >= MIN_BIN_THRE)
    {
        vector<uint64> vecBinCoors(binCoors.begin(), binCoors.end());
        std::random_shuffle(vecBinCoors.begin(), vecBinCoors.end());
        binCoors.clear();
        binCoors.insert(vecBinCoors.begin(), vecBinCoors.begin()+sampleNum);
#ifdef DEBUG
        cout<<"gef bin  coordinates number: "<<vecBinCoors.size()<<endl;
#endif
    }

    tissueSampleCoors.clear();
    for (auto& coor : tissueRawCoors)
    {
        uint32 y = coor & 0xFFFFFFFF;
        uint32 x = (coor >> 32) & 0xFFFFFFFF;
        uint64 binCoor = ((uint64)x/BIN_SIZE << 32) + y/BIN_SIZE;
        if (binCoors.count(binCoor) != 0)
            tissueSampleCoors.insert(coor);
    }

#ifdef DEBUG
    cout<<"gef raw  coordinates number: "<<tissueRawCoors.size()<<endl;
    cout<<"gef uniq coordinates number: "<<tissueSampleCoors.size()<<endl;
#endif

    return true;
}

uint64 extractExp(string expFile, unordered_set<uint64>& tissueSampleCoors, vector<ExpLine>& data)
{
    uint64 totalReads = 0;

    // Filter raw data by uniq coordinates after tissuecut
    auto readLine = [&](string& line) {
        uint64 cx, cy, gene, mid, cnt;
        libx::split(line, ' ', cy, cx, gene, mid, cnt);
        uint64 coor = ((uint64)cx << 32) + cy;
        if (tissueSampleCoors.count(coor) != 0)
        {
            ExpLine t(cx, cy, gene, mid);
            data.insert(data.end(), cnt, t);
        }
        totalReads += cnt;
        return true;
    };
    libx::readFile(expFile, readLine, "y");

#ifdef DEBUG
    cout<<"filtered and sampled gene expression lines: "<<data.size()<<endl;
    cout<<"total reads under tissue: "<<totalReads<<endl;
#endif

    // Check is there exists data after filtered by tissue.gef
    if (data.empty())
    {
        save_error_code(4, "no data left after filter by coordinates.");
        exit(-1);
    }

    return totalReads;
}

uint32 calcMedian(vector<uint32>& vec)
{
    if (vec.empty()) return 0;

    const auto midIter = vec.begin() + vec.size() / 2;
    std::nth_element(vec.begin(), midIter, vec.end());
    if (vec.size() % 2 == 0)
    {
        const auto& leftMidIter = std::max_element(vec.begin(),
            midIter);
        return (*leftMidIter + * midIter) / 2;
    }
    else
    {
        return *midIter;
    }
}

bool stat(SatStatMap& statMap, uint64& total, uint64& uniq, uint64& med)
{
    uint64 _total = 0, _uniq = 0;
    vector<uint32> geneNumList;
    unordered_set<uint32> uniqGeneSet;

    // structure: {Coor : {(Gene,MID) : cnt}}
    for (auto& [_, p] : statMap)
    {
        uniqGeneSet.clear();
        for (auto& [t, cnt] : p)
        {
            _total += cnt;
            uniqGeneSet.insert(std::get<0>(t));
        }
        _uniq += p.size();
        geneNumList.push_back(uniqGeneSet.size());
    }

    total = _total;
    uniq = _uniq;
    med = calcMedian(geneNumList);

    return true;
}

bool sample(vector<ExpLine>& data, uint64 totalReads, string satFile)
{
    SatStatMap rawStat;
    SatStatMap binStat;
    MIDStatMap midStat;
    stringstream ss;
    ss << "sample\tbar_x\tbar_y1\tbar_y2\tbar_umi\tbin_x\tbin_y1\tbin_y2\tbin_umi\n";
    
    std::random_shuffle(data.begin(), data.end());
    uint64 pos = 0;
    libx::Timer timer;
    int len = sizeof(SAMPLE_RATIOS) / sizeof(float);
    for (int i = 0; i < len; ++i)
    {
        float ratio = SAMPLE_RATIOS[i];
        ss<<ratio;
        uint64 endPos = data.size() * ratio;
        if (i == len-1)
            endPos = data.size();

        while (pos++ < endPos)
        {
            auto& t = data[pos];
            GeneMID gm(t.gene, t.mid);

            // collect raw data(bin1)
            uint64 coor = ((uint64)t.cx << 32) + t.cy;
            rawStat[coor][gm]++;

            // collect bin data(bin200)
            uint64 binCoor = ((uint64)t.cx/BIN_SIZE << 32) + t.cy/BIN_SIZE;
            binStat[binCoor][gm]++;

            // collect unique mid in bin200
            CoorGeneMID cgm(t.cx, t.cy, t.gene, t.mid);
            midStat[binCoor].insert(cgm);
        }

        uint64 realReads = totalReads * ratio;
        if (i == len-1)
            realReads = totalReads;
        // synchronize
        uint64 total, uniq, geMed;
        float sat;
        stat(rawStat, total, uniq, geMed);
        sat = total > 0 ? 1-uniq*1.0/total : 0;
        uint64 realUniq = total > 0 ? uniq * realReads / total : 0;
        ss<<"\t"<<realReads<<"\t"<<sat<<"\t"<<geMed<<"\t"<<realUniq;

        stat(binStat, total, uniq, geMed);
        sat = total > 0 ? 1-uniq*1.0/total : 0;
        uint64 sumCnt = 0;
        for (auto& [_, uniqSet] : midStat)
            sumCnt += uniqSet.size();
        uint64 aveCnt = midStat.empty() ? 0 : sumCnt / midStat.size();
        ss<<"\t"<<realReads<<"\t"<<sat<<"\t"<<geMed<<"\t"<<aveCnt;
        ss<<endl;

#ifdef DEBUG
        cout<<"ratio: "<<ratio<<" reads: "<<endPos<<" timer(s): "<<timer.toc()<<endl;
#endif
    }

    libx::writeFile(ss.str(), satFile);

    return true;
}

bool saturation(string tissueGefFile, string expFile, string satFile)
{
    libx::Timer timer;
    vector<ExpLine> data;
    uint64 totalReads;

    {
        unordered_set<uint64> tissueSampleCoors;
        extractTissueCoor(tissueGefFile, tissueSampleCoors);

#ifdef DEBUG
        cout<<"time(s) of read tissue.gef: "<<timer.toc()<<endl;
#endif

        totalReads = extractExp(expFile, tissueSampleCoors, data);

#ifdef DEBUG
        cout<<"time(s) of read raw gene exp file: "<<timer.toc()<<endl;
        // cout<<"maximum memory cost: "<<libx::getSelfMemory()<<endl;
#endif
    }

    sample(data, totalReads, satFile);

#ifdef DEBUG
    cout<<"time(s) of calculation: "<<timer.toc()<<endl;
#endif

    return true;
}

bool plot(string satFile, float ratio, string outDir)
{
    libx::Timer timer;
    fs::path scriptPath = binPath / "plot.pyc";
    std::string cmd("python3 -W ignore " + scriptPath.string() + " " + satFile +
                     " " + to_string(ratio) + " " + outDir + " 2>&1");
#ifdef DEBUG
    cout<<"run plot command: "<<cmd<<endl;
#endif

    std::string out;
    int rtn = libx::subprocess(cmd, out);
    if (rtn != 0)
    {
        cerr<<out<<endl;
        return false;
    }

#ifdef DEBUG
    cout<<"time(s) of plot: "<<timer.toc()<<endl;
#endif

    return true;
}

float prepare(vector<string>& annoFiles, vector<string>& mapFiles)
{
    // Extract the thrid value(annotated reads number) of thrid line in annotation stat file
    uint64 annoRead = 0;
    auto parseAnnoFile = [&](string& line) {
        vector<string> temp;
        libx::split(line, '\t', temp);
        annoRead = std::stoul(temp[2]);
        return false;
    };
    uint64 totalAnnoRead = 0;
    for (auto& annoFile : annoFiles)
    {
        annoRead = 0;
        libx::readFile(annoFile, parseAnnoFile, "#", 1);
        totalAnnoRead += annoRead;
    }

    // Extract the total reads in bcStar stat file
    uint64 mapRead = 0;
    auto parseMapFile = [&](string& line) {
        if (libx::startswith(line, "total_reads"))
        {
            vector<string> temp;
            libx::split(line, '\t', temp);
            mapRead = std::stoul(temp[1]);
            return false;
        }
        else
        {
            return true;
        }
    };
    uint64 totalMapRead = 0;
    for (auto& mapFile : mapFiles)
    {
        mapRead = 0;
        libx::readFile(mapFile, parseMapFile);
        totalMapRead += mapRead;
    }

    if (totalMapRead == 0)
    {
        save_error_code(5, "total map reads is 0, please check file format from --bcstat");
        exit(-1);
    }
    else if (totalMapRead < totalAnnoRead)
    {
        save_error_code(6, "map reads less than annotated reads.");
        exit(-1);
    }

    float res = totalAnnoRead * 1.0 / totalMapRead;
    
#ifdef DEBUG
    cout<<"anno reads: "<<totalAnnoRead<<" map reads: "<<totalMapRead<<" ratio: "<<res<<endl;
#endif

    return res;
}

int main(int argc, char** argv)
{
    // Parse arguments
    if (!parseArgs(argc, argv))
    {
        // cerr<<"Fail to parse arguments!"<<endl;
        return -1;
    }
    
    // Prepare
    float ratio = prepare(arguments.annoStatFiles, arguments.bcMapStatFiles);

    // Generate saturation file
    if (!saturation(arguments.tissueFile, arguments.expFile, arguments.outSatFile))
    {
        save_error_code(8, "fail to generate saturation file.");
        exit(-1);
    }

    // Plot
    if (!plot(arguments.outSatFile, ratio, arguments.outDir))
    {
        save_error_code(7, "plot error.");
        exit(-1);
    }

    return 0;
}
